### ISSUE TITLE
Add parameter to control API memory usage

### ISSUE THREAD

**Issue by AdriiiPRodri** [wazuh/wazuh#9055]
Hi team,

After the research done on the issue: #8876. We need to add a new parameter to the API to control its memory usage (maximum amount of RAM that the API can use).

It is a good practice to set some kind of limitation on the resource usage on the server-side. To accomplish this we will add a new parameter that states the maximum amount of memory that the API is capable of use.

This parameter is especially useful to mitigate the possible effects on the server in case of some kind of denial of service attack or vulnerability exploitation.

New tests should be added in Wazuh QA for this new API configuration. 

### wazuh/wazuh
- Unit tests without failures. Updated and/or expanded if there are new functions/methods/outputs:
  - [x] Cluster (`framework/wazuh/core/cluster/tests/` & `framework/wazuh/core/cluster/dapi/tests/`)
  - [x] Core (`framework/wazuh/core/tests/`)
  - [x] SDK (`framework/wazuh/tests/`)
  - [x] RBAC (`framework/wazuh/rbac/tests/`)
  - [x] API (`api/api/tests/`)
- API tavern integration tests without failures. Updated and/or expanded if needed (`api/test/integration/`):
  - [x] Affected tests 
  - [x] Affected RBAC (black and white) tests
- [x] Review integration test mapping using the script (`api/test/integration/mapping/integration_test_api_endpoints.json`)
- [x] Review of spec.yaml examples and schemas (`api/api/spec/spec.yaml`)
- [x] Review exceptions remediation when any endpoint path changes or is removed (`framework/wazuh/core/exception.py`)
- [x] Changelog (`CHANGELOG.md`)

### wazuh/wazuh-documentation
- [x] Migration from 3.X for changed endpoints (`source/user-manual/api/equivalence.rst`)
- [x] Update RBAC reference with new/modified actions/resources/relationships (`source/user-manual/api/rbac/reference.rst`)

**Comment by mcarmona99** [comment]
## Issue update


This limit can be combined with the API timeout so requests are cancelled for certain conditions:

- Timeout + Max RAM: Request is canceled when either of these two things happens before (timeout or exceeded RAM usage)

- Only max RAM: Request is canceled if the response exceeds the maximum size allowed, regardless of the time required for it.

- Only timeout: Current behavior. The request is canceled only if the timeout is exceeded.


With the following command, we can see the API memory consumption in real time (percentage)

`watch -n0 'ps -O %mem -A | grep api'
`

`   4911  0.5 S ?        00:00:07 /var/ossec/framework/python/bin/python3 /var/ossec/api/scripts/wazuh-apid.py
`

`ps aux` also shows `%mem` and `%cpu`, but we need the exact amount. To see this RAM consumed, we can see the `/proc/<pid>/status` file.

The most important field is the `VmSize`, which should be close to the percentage `ps aux` gives.

```
root@wazuh-master:/# cat /proc/4911/status | grep VmSize
VmSize:	  280692 kB
```

In this case, wazuh-apid is consuming `280,692 mB`. I used this command after an API restart.

After some request to the MITRE endpoints (that use a cache to save information), the memory consumed is `429,980 mB`. 



There's no utility to change the limit for a running process (it exists but it is a builtint Linux command for Linux system with specific kernels)

We can limit the memory of new processes. After that, we can create the wazuh-apid process and then we can revert this limit again.

`ulimit`: Provides control over the resources available to the shell and processes it creates, on systems that allow such control. 

`ulimit -v 1024
` 

1024 are kBs

`-v` indicates the virtual memory limit for new processes created.
Values are in 1024-byte increments, except for `-t`, which is in seconds, `-p`, which is in increments of 512 bytes, and `-u`, which is an unscaled number of processes.


This sets the memory usage limit for the shell and all its descendants.

We can use this when creating the apid process.

After that, we set the limit value to the original value.


I have tried `ulimit` in an example and we cannot limit the memory usage inside the `wazuh-apid.py` as the limitation must follow this format:

`ulimit -v 12200 && python3 script.py
`

In this example, `script.py` will never use more than 12200 kB:

`root      130428  0.0  0.0  12136  7960 pts/0    S+   16:12   0:00 python3 script.py
`

I have tested this and without the limitation, the script uses 12300 kB aprox.

If the maximum memory assigned if too litle, the following error appears:

```
root@wazuh-master:/# ulimit -v 12100 && python3 script.py
Traceback (most recent call last):
  File "script.py", line 8, in <module>
  File "/usr/lib/python3.8/copy.py", line 52, in <module>
  File "/usr/lib/python3.8/weakref.py", line 90, in <module>
  File "/usr/lib/python3.8/abc.py", line 85, in __new__
MemoryError
```


I keep investigating how to limit the wazuh api memory usage.

**Comment by mcarmona99** [comment]
## Issue update

We can use the following code to restrict the `wazuh-apid.py` memory usage.

The code puts a limit on te total address space.

`RLIMIT_AS` = The maximum area (in bytes) of address space which may be taken by the process.

```
# using resource 
import resource
  
def limit_memory(maxsize):
    soft, hard = resource.getrlimit(resource.RLIMIT_AS)
    resource.setrlimit(resource.RLIMIT_AS, (maxsize, hard))
```
    
When no more memory is available then the program will start generating `MemoryError` exceptions with a memory limit.


Script to test this:

```
import resource

import numpy as np


def limit_memory(maxsize):
    # maxsize must be bytes
    soft, hard = resource.getrlimit(resource.RLIMIT_AS)
    print("Limit starts as:", soft, hard)
    resource.setrlimit(resource.RLIMIT_AS, (maxsize, hard))
    soft, hard = resource.getrlimit(resource.RLIMIT_AS)
    print("Limit is now:", soft, hard, "\n")


def peak_virtual_memory_mb():
    with open('/proc/self/status') as f:
        status = f.readlines()
        vmpeak = next(s for s in status if s.startswith("VmPeak:"))
        return vmpeak


# Set limit to 1 GB
limit_memory(int(1e9))

print(peak_virtual_memory_mb())

print("Allocating 80 KB, should work")
M1 = np.arange(100 * 100, dtype="u8")

print(peak_virtual_memory_mb())

print("Allocating 80 MB, should work")
M2 = np.arange(1000 * 1000 * 10, dtype="u8")

print(peak_virtual_memory_mb())

print("Allocating 2 GB, should fail")
M3 = np.arange(1000 * 1000 * 250, dtype="u8")

print(peak_virtual_memory_mb())

input("Still here…")
```


I have added the `peak_virtual_memory_mb` function to check virtual memory usage at any point.






This is the output of the script:

```
python3 /scratches/scratch_10.py 
Limit starts as: -1 -1
Limit is now: 1000000000 -1 

VmPeak:	  680852 kB

Allocating 80 KB, should work
VmPeak:	  680852 kB

Allocating 80 MB, should work
VmPeak:	  704840 kB

Allocating 2 GB, should fail
Traceback (most recent call last):
  File "/scratches/scratch_10.py", line 38, in <module>
    M3 = np.arange(1000 * 1000 * 250, dtype="u8")
numpy.core._exceptions.MemoryError: Unable to allocate 1.86 GiB for an array with shape (250000000,) and data type uint64
```


In this case the MemoryError is a numpy exception.
If we test this with a Python list instead of using numpy, the exception will be a generic MemoryError:

```
Traceback (most recent call last):
  File "/scratches/scratch_10.py", line 38, in <module>
    M3 = [1]*213909000000
MemoryError
```


This is what I am going to implement in our `wazuh-apid` script, the memory usage parameter will be available in our `api.yaml` configuration file.

Regards,
Manuel.

**Comment by mcarmona99** [comment]
## Issue update

What happens if this is used with a `processpoolexecutor`?? I am going to check if child processes have the memory limitation. In case they have it, we can say the limitation is in terms of processes and childs inherit this limit.


With the following code, we can see if we create childs, the address space limit is the same as the parent process one:

<details>

<summary>scratch.py</summary>

```
import asyncio
import os
import random
import resource
from concurrent.futures import ProcessPoolExecutor
from time import sleep


def limit_memory(maxsize):
    # maxsize must be bytes
    soft, hard = resource.getrlimit(resource.RLIMIT_AS)
    print("Limit starts as:", soft, hard)
    resource.setrlimit(resource.RLIMIT_AS, (maxsize, hard))
    soft, hard = resource.getrlimit(resource.RLIMIT_AS)
    print("Limit is now:", soft, hard, "\n")


processpool = ProcessPoolExecutor()


def task():
    sleep(random.randint(1, 2))
    soft, hard = resource.getrlimit(resource.RLIMIT_AS)
    print('Executing our Task on process {}'.format(os.getpid()),
          "Limit is now:", soft, hard)
    sleep(random.randint(1, 2))


async def main():
    limit_memory(int(1e9))

    loop = asyncio.get_event_loop()
    task1 = loop.run_in_executor(processpool, task)
    task2 = loop.run_in_executor(processpool, task)
    task3 = loop.run_in_executor(processpool, task)

    await asyncio.wait_for(task1, timeout=15)
    await asyncio.wait_for(task2, timeout=15)
    await asyncio.wait_for(task3, timeout=15)


asyncio.run(main())
```

</details>

Note: we are only changing the soft limit as this is the value Linux uses to limit the system resources for running processes. The hard limit is the maximum value that is allowed for the soft limit.


By default, the API maximum memory usage will be 1,024 GB. The parameter will be located in the access configuration block of the `api.yaml`.

```
# access:
#  ...
#  max_memory_usage: 1024000000
```

I have been testing with different maximum memory usage values to see the errors raised when we exceed the limit.


With this `max_memory_usage`, if we exceed the limit in the API requests, this will apppear in our `api.log`:

```
Traceback (most recent call last):
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.3.0-py3.9.egg/wazuh/core/cluster/dapi/dapi.py", line 254, in execute_local_request
    data = await asyncio.wait_for(task, timeout=timeout)
  File "/var/ossec/framework/python/lib/python3.9/asyncio/tasks.py", line 481, in wait_for
    return fut.result()
  File "/var/ossec/framework/python/lib/python3.9/concurrent/futures/thread.py", line 52, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.3.0-py3.9.egg/wazuh/core/cluster/dapi/dapi.py", line 232, in run_local
    data = self.f(**self.f_kwargs)
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.3.0-py3.9.egg/wazuh/rbac/decorators.py", line 406, in wrapper
    result = func(*args, **kwargs) if not skip_execution else None
  File "/var/ossec/framework/python/lib/python3.9/site-packages/wazuh-4.3.0-py3.9.egg/wazuh/mitre.py", line 111, in mitre_references
    a = [None]*3132113131132
MemoryError
2021/07/09 10:34:58 INFO: wazuh 172.19.0.1 "GET /mitre/references" with parameters {"reference_ids": "course-of-action--00d7d21b-69d6-4797-88a2-c86f3fc97651"} and body {} done in 0.208s: 500
```

We also need to handle this exception.

To sum up, we can add a parameter to limit the max memory that can be allocated with the `wazuh-apid` process. 


**Note**:
When setting the process limit, we can use `RLIMIT_VMEM` and `RLIMIT_AS`. What are the differencies between these 2 variables?

The `resource` module does not attempt to mask platform differences — symbols not defined for a platform will not be available from this module on that platform.
According to the bash `ulimit` command, it uses `RLIMIT_AS` if `RLIMIT_VMEM` is not defined.

In other words, when `RLIMIT_VMEM` is defined, it is an alias for `RLIMIT_AS`.

Regards,
Manuel.

**Comment by davidjiglesias** [comment]
After the investigation performed in this issue we have decided not to implement the functionality.
